pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol

```diff
@@ -0,0 +1,299 @@
+ * all token balances go down together, the *relative* change would be zero, and the breaker would not be
+ * triggered: even though the external price might have dropped 50 or 70%. It is only the *relative* movement
+ * compared to the rest of the pool that matters.
+ *
+ * If we have tokens A, B, and C, If A drops 20% and B and C are unchanged, that's a simple 20% drop for A.
+ * However, if A is unchanged and C increases 25%, that would also be a 20% "drop" for A 1 / 1.25 = 0.8.
+ * The breaker might register a 20% drop even if both go up - if our target token lags the market. For
+ * instance, if A goes up 60% and B and C double, 1.6 / 2 = 0.8.
+ *
+ * Since BPT prices are not intuitive - and there is a very non-linear relationship between "spot" prices and
+ * BPT prices - circuit breakers are set using simple percentages. Intuitively, a lower bound of 0.8 means the
+ * token can lose 20% of its value before triggering the circuit breaker, and an upper bound of 3.0 means it
+ * can triple before being halted. These percentages are then transformed into BPT prices for comparison to the
+ * "reference" state of the pool when the circuit breaker was set.
+ *
+ * Prices can change in two ways: arbitrage traders responding to external price movement can change the balances,
+ * or an ongoing gradual weight update (or change in pool composition) can change the weights. In order to isolate
+ * the balance changes due to price movement, the bounds are dynamic, adjusted for the current weight.
+ */
+library CircuitBreakerLib {
+    using ValueCompression for uint256;
+    using FixedPoint for uint256;
+    using WordCodec for bytes32;
+
+    struct CircuitBreakerParams {
+        uint256 bptPrice;
+        uint256 weightComplement;
+        uint256 lowerBound;
+        uint256 upperBound;
+    }
+
+    // Store circuit breaker information per token
+    // When the circuit breaker is set, the caller passes in the lower and upper bounds (expressed as percentages),
+    // and the current weight complement (1 - weight). Since this value is bounded by 1e18, which fits in ~60 bits,
+    // there is no need for compression.
+    //
+    // We then compute and store the current BPT price, and the lower and upper bound conversion ratios, used to
+    // convert the percentage bounds into BPT prices that can be directly compared to the "runtime" BPT prices.
+    //
+    // Since the price bounds shift along with the token weight, in general these bound ratios would need to be
+    // computed every time. However, if the weight of the token and composition of the pool have not changed since
+    // the circuit breaker was set, these stored values can still be used, avoiding a heavy computation.
+    //
+    // [      24 bits      |      24 bits      |  112 bits |   64 bits    |   16 bits   |   16 bits   |
+    // [ upper bound ratio | lower bound ratio | BPT price | weight comp. | upper bound | lower bound |
```

Comment:
```suggestion
    // [      24 bits      |      24 bits      |  112 bits |   60 bits    |   18 bits   |   18 bits   |
    // [ upper bound ratio | lower bound ratio | BPT price | weight comp. | upper bound | lower bound |
```

As mentioned, in the comment the weight complement fits in 60 bits so we can get some extra precision on the bounds for free like this.

Reply:
Ah, just saw you mentioned this in slack.

Reply:
What about the BPT price? It's now 112 bits (max Vault balance). If that could be, say, 96 bits (~10^10) we'd get a bunch more. I also think any extra precision might be better assigned to the bound ratios - those are the most sensitive. Presumably the bounds will be set to "round" numbers like 0.8 or 0.9 anyway.

I have another version that removes compression from bounds entirely.

Reply:
huh, I'm a bit confused about naming here in that case.

Working off of the name of this constant
https://github.com/balancer-labs/balancer-v2-monorepo/blob/ce65e40000d8f8b2cea132eb8483e0d7b00e5aaf/pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol#L105
along with these lines
https://github.com/balancer-labs/balancer-v2-monorepo/blob/ce65e40000d8f8b2cea132eb8483e0d7b00e5aaf/pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol#L120-L127

I was assuming that we're only compressing the bounds, and so that's where we need extra bits.

It looks like we're compressing the bound ratios as well.
https://github.com/balancer-labs/balancer-v2-monorepo/blob/ce65e40000d8f8b2cea132eb8483e0d7b00e5aaf/pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol#L277-L287

In this case I think we should add a `_MAX_BOUND_RATIO` constant (even if it's just `uint256 private constant _MAX_BOUND_RATIO = _MAX_BOUND;`) as it isn't immediately obvious that these values should be compressed to the same range.